File: python/sobec/walk/ocp.py
import sys

import pinocchio as pin
import crocoddyl as croc
import numpy as np


# Local imports
import sobec
from .weight_share import computeReferenceForces


# workaround python 2
if sys.version_info.major < 3:
    FileNotFoundError = IOError


def buildRunningModels(robotWrapper, contactPattern, params):

    p = params
    robot = robotWrapper

    referenceForces = computeReferenceForces(
        contactPattern, robot.gravForce, params.transitionDuration
    )
    models = []

    # #################################################################################
    for t, pattern in enumerate(contactPattern[:-1]):
        # print("time t=%s %s" % (t, pattern))

        # Basics
        state = croc.StateMultibody(robot.model)
        actuation = croc.ActuationModelFloatingBase(state)

        # Contacts
        contacts = croc.ContactModelMultiple(state, actuation.nu)
        for k, cid in enumerate(robot.contactIds):
            if not pattern[k]:
                continue
            contact = croc.ContactModel6D(
                state, cid, pin.SE3.Identity(), actuation.nu, p.baumgartGains
            )
            contacts.addContact(robot.model.frames[cid].name + "_contact", contact)

        # Costs
        costs = croc.CostModelSum(state, actuation.nu)

        xRegResidual = croc.ResidualModelState(state, robot.x0, actuation.nu)
        xRegCost = croc.CostModelResidual(
            state,
            croc.ActivationModelWeightedQuad(p.stateImportance**2),
            xRegResidual,
        )
        if p.refStateWeight > 0:
            costs.addCost("stateReg", xRegCost, p.refStateWeight)

        uResidual = croc.ResidualModelControl(state, actuation.nu)
        uRegCost = croc.CostModelResidual(
            state,
            croc.ActivationModelWeightedQuad(np.array(p.controlImportance**2)),
            uResidual,
        )
        if p.refTorqueWeight > 0:
            costs.addCost("ctrlReg", uRegCost, p.refTorqueWeight)

        comResidual = croc.ResidualModelCoMPosition(state, robot.com0, actuation.nu)
        comAct = croc.ActivationModelWeightedQuad(np.array([0, 0, 1]))
        comCost = croc.CostModelResidual(state, comAct, comResidual)
        if p.comWeight > 0:
            costs.addCost("com", comCost, p.comWeight)

        comVelResidual = sobec.ResidualModelCoMVelocity(state, p.vcomRef, actuation.nu)
        comVelAct = croc.ActivationModelWeightedQuad(p.vcomImportance)
        comVelCost = croc.CostModelResidual(state, comVelAct, comVelResidual)
        if p.vcomWeight > 0:
            costs.addCost("comVelCost", comVelCost, p.vcomWeight)

        # Contact costs
        for k, cid in enumerate(robot.contactIds):
            if not pattern[k]:
                continue

            copResidual = sobec.ResidualModelCenterOfPressure(state, cid, actuation.nu)
            copAct = croc.ActivationModelWeightedQuad(
                np.array([1.0 / p.footSize**2] * 2)
            )
            copCost = croc.CostModelResidual(state, copAct, copResidual)
            if p.copWeight > 0:
                costs.addCost(
                    "%s_cop" % robot.model.frames[cid].name, copCost, p.copWeight
                )

            # Cone with enormous friction (Assuming the robot will barely ever slide).
            # p.footSize is the allowed area size, while cone expects the corner
            # coordinates => x2
            cone = croc.WrenchCone(
                np.eye(3), 1000, np.array([p.footSize * 2] * 2), 4, True, 1, 10000
            )
            coneCost = croc.ResidualModelContactWrenchCone(
                state, cid, cone, actuation.nu
            )
            ub = cone.ub.copy()
            ub[:4] = np.inf
            # ub[5:] = np.inf  ### DEBUG
            ub[-8:] = np.inf
            coneAct = croc.ActivationModelQuadraticBarrier(
                croc.ActivationBounds(cone.lb, ub)
            )
            coneCost = croc.CostModelResidual(state, coneAct, coneCost)
            if p.conePenaltyWeight:
                costs.addCost(
                    "%s_cone" % robot.model.frames[cid].name,
                    coneCost,
                    p.conePenaltyWeight,
                )

            # Penalize the distance to the central axis of the cone ...
            #  ... using normalization weights depending on the axis.
            # The weights are squared to match the tuning of the CASADI formulation.
            coneAxisResidual = croc.ResidualModelContactForce(
                state, cid, pin.Force.Zero(), 6, actuation.nu
            )
            w = np.array(p.forceImportance**2)
            w[2] = 0
            coneAxisAct = croc.ActivationModelWeightedQuad(w)
            coneAxisCost = croc.CostModelResidual(state, coneAxisAct, coneAxisResidual)
            if p.coneAxisWeight > 0:
                costs.addCost(
                    "%s_coneaxis" % robot.model.frames[cid].name,
                    coneAxisCost,
                    p.coneAxisWeight,
                )

            # Follow reference (smooth) contact forces
            forceRefResidual = croc.ResidualModelContactForce(
                state, cid, pin.Force(referenceForces[t][k]), 6, actuation.nu
            )
            forceRefCost = croc.CostModelResidual(state, forceRefResidual)
            if p.refForceWeight > 0:
                costs.addCost(
                    "%s_forceref" % robot.model.frames[cid].name,
                    forceRefCost,
                    p.refForceWeight / robot.gravForce**2,
                )

        # IMPACT
        for k, cid in enumerate(robot.contactIds):
            if t > 0 and not contactPattern[t - 1][k] and pattern[k]:
                # REMEMBER TO divide the weight by p.DT, as impact should be independant
                # of the node duration (at least, that s how weights are tuned in
                # casadi).

                print("Impact %s at time %s" % (cid, t))
                impactResidual = croc.ResidualModelFrameTranslation(
                    state, cid, np.zeros(3), actuation.nu
                )
                impactAct = croc.ActivationModelWeightedQuad(np.array([0, 0, 1]))
                impactCost = croc.CostModelResidual(state, impactAct, impactResidual)
                if p.impactAltitudeWeight > 0:
                    costs.addCost(
                        "%s_altitudeimpact" % robot.model.frames[cid].name,
                        impactCost,
                        p.impactAltitudeWeight / p.DT,
                    )

                impactVelResidual = croc.ResidualModelFrameVelocity(
                    state,
                    cid,
                    pin.Motion.Zero(),
                    pin.ReferenceFrame.LOCAL,
                    actuation.nu,
                )
                impactVelCost = croc.CostModelResidual(state, impactVelResidual)
                if p.impactVelocityWeight > 0:
                    costs.addCost(
                        "%s_velimpact" % robot.model.frames[cid].name,
                        impactVelCost,
                        p.impactVelocityWeight / p.DT,
                    )

                impactRotResidual = croc.ResidualModelFrameRotation(
                    state, cid, np.eye(3), actuation.nu
                )
                impactRotAct = croc.ActivationModelWeightedQuad(np.array([1, 1, 0]))
                impactRotCost = croc.CostModelResidual(
                    state, impactRotAct, impactRotResidual
                )
                if p.impactRotationWeight > 0:
                    costs.addCost(
                        "%s_rotimpact" % robot.model.frames[cid].name,
                        impactRotCost,
                        p.impactRotationWeight / p.DT,
                    )

                impactRefJointsResidual = croc.ResidualModelState(
                    state, robot.x0, actuation.nu
                )
                jselec = np.zeros(robot.model.nv * 2)
                jselec[
                    [
                        robot.model.joints[robot.model.getJointId(name)].idx_v
                        for name in p.mainJointIds
                    ]
                ] = 1
                impactRefJointsAct = croc.ActivationModelWeightedQuad(jselec)
                impactRefJointCost = croc.CostModelResidual(
                    state, impactRefJointsAct, impactRefJointsResidual
                )
                if p.refMainJointsAtImpactWeight > 0:
                    costs.addCost(
                        "impactRefJoint",
                        impactRefJointCost,
                        p.refMainJointsAtImpactWeight / p.DT,
                    )

        # Flying foot
        for k, fid in enumerate(robot.contactIds):
            if pattern[k]:
                continue
            verticalFootVelResidual = croc.ResidualModelFrameVelocity(
                state,
                fid,
                pin.Motion.Zero(),
                pin.ReferenceFrame.LOCAL_WORLD_ALIGNED,
                actuation.nu,
            )
            verticalFootVelAct = croc.ActivationModelWeightedQuad(
                np.array([0, 0, 1, 0, 0, 0])
            )
            verticalFootVelCost = croc.CostModelResidual(
                state, verticalFootVelAct, verticalFootVelResidual
            )
            if p.verticalFootVelWeight > 0:
                costs.addCost(
                    "%s_vfoot_vel" % robot.model.frames[fid].name,
                    verticalFootVelCost,
                    p.verticalFootVelWeight,
                )

            # Slope is /2 since it is squared in casadi (je me comprends)
            flyHighResidual = sobec.ResidualModelFlyHigh(
                state, fid, p.flyHighSlope / 2.0, actuation.nu
            )
            flyHighCost = croc.CostModelResidual(state, flyHighResidual)
            if p.flyHighWeight > 0:
                costs.addCost(
                    "%s_flyhigh" % robot.model.frames[fid].name,
                    flyHighCost,
                    p.flyHighWeight,
                )

            groundColRes = croc.ResidualModelFrameTranslation(
                state, fid, np.zeros(3), actuation.nu
            )
            # groundColBounds = croc.ActivationBounds(
            # np.array([-np.inf, -np.inf, 0.01]), np.array([np.inf, np.inf, np.inf])
            # )
            # np.inf introduces an error on lb[2] ... why? TODO ... patch by replacing
            # np.inf with 1000
            groundColBounds = croc.ActivationBounds(
                np.array([-1000, -1000, 0.0]), np.array([1000, 1000, 1000])
            )
            groundColAct = croc.ActivationModelQuadraticBarrier(groundColBounds)
            groundColCost = croc.CostModelResidual(state, groundColAct, groundColRes)
            if p.groundColWeight > 0:
                costs.addCost(
                    "%s_groundcol" % robot.model.frames[fid].name,
                    groundColCost,
                    p.groundColWeight,
                )

            for kc, cid in enumerate(robot.contactIds):
                if not pattern[kc]:
                    continue
                assert fid != cid
                # print("At t=%s add collision between %s and %s" % (t, cid, fid))

                for id1, id2 in [
                    (i, j)
                    for i in [cid, robot.towIds[cid], robot.heelIds[cid]]
                    for j in [fid, robot.towIds[fid], robot.heelIds[fid]]
                ]:
                    feetColResidual = sobec.ResidualModelFeetCollision(
                        state, id1, id2, actuation.nu
                    )
                    feetColBounds = croc.ActivationBounds(
                        np.array([p.footMinimalDistance]), np.array([1000])
                    )
                    feetColAct = croc.ActivationModelQuadraticBarrier(feetColBounds)
                    feetColCost = croc.CostModelResidual(
                        state, feetColAct, feetColResidual
                    )
                    if p.feetCollisionWeight > 0:
                        costs.addCost(
                            (
                                "feetcol_%s_VS_%s"
                                % (
                                    robot.model.frames[id1].name,
                                    robot.model.frames[id2].name,
                                )
                            ),
                            feetColCost,
                            p.feetCollisionWeight,
                        )

        # Action
        damodel = croc.DifferentialActionModelContactFwdDynamics(
            state, actuation, contacts, costs, p.kktDamping, True
        )
        amodel = croc.IntegratedActionModelEuler(damodel, p.DT)

        models.append(amodel)

    return models


# ### TERMINAL MODEL ##################################################################
def buildTerminalModel(robotWrapper, contactPattern, params):

    robot = robotWrapper
    p = params
    pattern = contactPattern[-1]

    # Horizon length
    T = len(contactPattern) - 1

    state = croc.StateMultibody(robot.model)
    actuation = croc.ActuationModelFloatingBase(state)

    # Contacts
    contacts = croc.ContactModelMultiple(state, actuation.nu)
    for k, cid in enumerate(robot.contactIds):
        if not pattern[k]:
            continue
        contact = croc.ContactModel6D(
            state, cid, pin.SE3.Identity(), actuation.nu, p.baumgartGains
        )
        contacts.addContact(robot.model.frames[cid].name + "_contact", contact)

    # Costs
    costs = croc.CostModelSum(state, actuation.nu)

    # if "stateTerminalTarget" not in locals():
    stateTerminalTarget = robot.x0.copy()
    stateTerminalTarget[:3] += p.vcomRef * T * p.DT
    stateTerminalResidual = croc.ResidualModelState(
        state, stateTerminalTarget, actuation.nu
    )
    stateTerminalAct = croc.ActivationModelWeightedQuad(p.stateTerminalImportance**2)
    stateTerminalCost = croc.CostModelResidual(
        state, stateTerminalAct, stateTerminalResidual
    )
    if p.stateTerminalWeight > 0:
        costs.addCost("stateReg", stateTerminalCost, p.stateTerminalWeight)

    damodel = croc.DifferentialActionModelContactFwdDynamics(
        state, actuation, contacts, costs, p.kktDamping, True
    )
    termmodel = croc.IntegratedActionModelEuler(damodel, p.DT)

    return termmodel


# ### SOLVER ########################################################################


def buildSolver(robotWrapper, contactPattern, walkParams):
    models = buildRunningModels(robotWrapper, contactPattern, walkParams)
    termmodel = buildTerminalModel(robotWrapper, contactPattern, walkParams)

    problem = croc.ShootingProblem(robotWrapper.x0, models, termmodel)
    ddp = croc.SolverFDDP(problem)
    ddp.th_stop = walkParams.solver_th_stop
    return ddp


def buildInitialGuess(problem, walkParams):
    if walkParams.guessFile is not None:
        try:
            guess = np.load(walkParams.guessFile, allow_pickle=True)[()]
            print('Load "%s"!' % walkParams.guessFile)
            x0s = [x for x in guess["xs"]]
            u0s = [u for u in guess["us"]]
        except (FileNotFoundError, KeyError):
            x0s = []
            u0s = []
    else:
        x0s = []
        # us0 = []

    if len(x0s) != problem.T + 1 or len(u0s) != problem.T:
        print("No valid solution file, build quasistatic initial guess!")
        x0s = [problem.x0.copy() for _ in range(problem.T + 1)]
        u0s = [
            m.quasiStatic(d, x)
            for m, d, x in zip(problem.runningModels, problem.runningDatas, x0s)
        ]

    return x0s, u0s


# ### SOLUTION ######################################################################


class Solution:
    def __init__(self, robotWrapper, ddp):
        # model = ddp.problem.terminalModel.differential.pinocchio
        self.xs = np.array(ddp.xs)
        self.us = np.array(ddp.us)
        self.acs = np.array([d.differential.xout for d in ddp.problem.runningDatas])
        self.fs = [
            [
                (cd.data().jMf.inverse() * cd.data().f).vector
                for cd in d.differential.multibody.contacts.contacts
            ]
            for d in ddp.problem.runningDatas
        ]
        self.fs0 = [
            np.concatenate(
                [
                    (
                        d.differential.multibody.contacts.contacts[
                            "%s_contact" % robotWrapper.model.frames[cid].name
                        ].jMf.inverse()
                        * d.differential.multibody.contacts.contacts[
                            "%s_contact" % robotWrapper.model.frames[cid].name
                        ].f
                    ).vector
                    if "%s_contact" % robotWrapper.model.frames[cid].name
                    in d.differential.multibody.contacts.contacts
                    else np.zeros(6)
                    for cid in robotWrapper.contactIds
                ]
            )
            for m, d in zip(ddp.problem.runningModels, ddp.problem.runningDatas)
        ]

File: python/walk-without-think/ocp.cpp
///////////////////////////////////////////////////////////////////////////////
// BSD 3-Clause License
//
// Copyright (C) 2022, LAAS-CNRS
// Copyright note valid unless otherwise stated in individual files.
// All rights reserved.
///////////////////////////////////////////////////////////////////////////////

#include "sobec/walk-without-think/ocp.hpp"

#include <pinocchio/fwd.hpp>

#include <boost/python.hpp>
#include <boost/python/enum.hpp>
#include <crocoddyl/core/solvers/fddp.hpp>
#include <eigenpy/eigenpy.hpp>
#include <pinocchio/multibody/fwd.hpp>

#include "sobec/fwd.hpp"

namespace sobec {
namespace python {

using namespace crocoddyl;
namespace bp = boost::python;

void exposeOCPRobotWrapper() {
  bp::register_ptr_to_python<boost::shared_ptr<OCPRobotWrapper> >();

  bp::class_<OCPRobotWrapper>(
      "OCPRobotWrapper",
      bp::init<boost::shared_ptr<pinocchio::Model>, std::string, std::string>(
          bp::args("self", "model", "contactKey", "reference_posture"), "Initialize the OCP from robot and params"))
      .add_property("x0", bp::make_getter(&OCPRobotWrapper::x0, bp::return_internal_reference<>()),
                    bp::make_setter(&OCPRobotWrapper::x0), "Reference state")
      .add_property("com0", bp::make_getter(&OCPRobotWrapper::com0, bp::return_internal_reference<>()),
                    bp::make_setter(&OCPRobotWrapper::com0), "Reference com value (computed at x0)")
      .add_property("model",
                    bp::make_getter(&OCPRobotWrapper::model,
                                    // bp::return_internal_reference<>()),
                                    bp::return_value_policy<bp::return_by_value>()),
                    bp::make_setter(&OCPRobotWrapper::model), "Pinocchio model")
      .add_property("data", bp::make_getter(&OCPRobotWrapper::data, bp::return_internal_reference<>()),
                    "pinocchio data")
      .add_property("contactIds", bp::make_getter(&OCPRobotWrapper::contactIds, bp::return_internal_reference<>()),
                    bp::make_setter(&OCPRobotWrapper::contactIds),
                    "List of the end-effector frame Id that can be in contact.")
      .add_property("towIds", bp::make_getter(&OCPRobotWrapper::towIds, bp::return_internal_reference<>()),
                    bp::make_setter(&OCPRobotWrapper::towIds), "TODO?")
      .add_property("heelIds", bp::make_getter(&OCPRobotWrapper::heelIds, bp::return_internal_reference<>()),
                    bp::make_setter(&OCPRobotWrapper::heelIds), "TODO?");
}
void exposeOCPParams() {
  bp::register_ptr_to_python<boost::shared_ptr<OCPWalkParams> >();

  bp::class_<OCPWalkParams>("OCPWalkParams", bp::init<>(bp::args("self"), "Empty initialization of the OCP params"))
      // .add_property(
      //     "solver_reg_min", bp::make_getter(&OCPWalkParams::solver_reg_min),
      //     bp::make_setter(&OCPWalkParams::solver_reg_min),
      //     "reg_min param (minimal regularization) to configure the solver.")
      // .add_property("solver_maxiter",
      // bp::make_getter(&OCPWalkParams::solver_maxiter),
      //               bp::make_setter(&OCPWalkParams::solver_maxiter),
      //               "maxiter param to configure the solver.")
      .def("readFromYaml", &OCPWalkParams::readParamsFromYamlFile, bp::args("filename"))
      .add_property("DT", bp::make_getter(&OCPWalkParams::DT), bp::make_setter(&OCPWalkParams::DT),
                    "time step duration of the shooting nodes.")

      .add_property("mainJointIds", bp::make_getter(&OCPWalkParams::mainJointIds),
                    bp::make_setter(&OCPWalkParams::mainJointIds), "Most important (big weight) joints at impact.")
      .add_property("baumgartGains", bp::make_getter(&OCPWalkParams::baumgartGains, bp::return_internal_reference<>()),
                    bp::make_setter(&OCPWalkParams::baumgartGains), ".")

      .add_property("stateImportance",
                    bp::make_getter(&OCPWalkParams::stateImportance, bp::return_internal_reference<>()),
                    bp::make_setter(&OCPWalkParams::stateImportance), ".")

      .add_property("stateTerminalImportance",
                    bp::make_getter(&OCPWalkParams::stateTerminalImportance, bp::return_internal_reference<>()),
                    bp::make_setter(&OCPWalkParams::stateTerminalImportance), ".")

      .add_property("controlImportance",
                    bp::make_getter(&OCPWalkParams::controlImportance, bp::return_internal_reference<>()),
                    bp::make_setter(&OCPWalkParams::controlImportance), ".")

      .add_property("vcomImportance",
                    bp::make_getter(&OCPWalkParams::vcomImportance, bp::return_internal_reference<>()),
                    bp::make_setter(&OCPWalkParams::vcomImportance), ".")

      .add_property("forceImportance",
                    bp::make_getter(&OCPWalkParams::forceImportance, bp::return_internal_reference<>()),
                    bp::make_setter(&OCPWalkParams::forceImportance), ".")

      .add_property("vcomRef", bp::make_getter(&OCPWalkParams::vcomRef, bp::return_internal_reference<>()),
                    bp::make_setter(&OCPWalkParams::vcomRef), ".")

      .add_property("footSize", bp::make_getter(&OCPWalkParams::footSize), bp::make_setter(&OCPWalkParams::footSize),
                    ".")
      .add_property("refStateWeight", bp::make_getter(&OCPWalkParams::refStateWeight),
                    bp::make_setter(&OCPWalkParams::refStateWeight), ".")
      .add_property("refTorqueWeight", bp::make_getter(&OCPWalkParams::refTorqueWeight),
                    bp::make_setter(&OCPWalkParams::refTorqueWeight), ".")
      .add_property("comWeight", bp::make_getter(&OCPWalkParams::comWeight),
                    bp::make_setter(&OCPWalkParams::comWeight), ".")
      .add_property("vcomWeight", bp::make_getter(&OCPWalkParams::vcomWeight),
                    bp::make_setter(&OCPWalkParams::vcomWeight), ".")
      .add_property("copWeight", bp::make_getter(&OCPWalkParams::copWeight),
                    bp::make_setter(&OCPWalkParams::copWeight), ".")
      .add_property("conePenaltyWeight", bp::make_getter(&OCPWalkParams::conePenaltyWeight),
                    bp::make_setter(&OCPWalkParams::conePenaltyWeight), ".")
      .add_property("coneAxisWeight", bp::make_getter(&OCPWalkParams::coneAxisWeight),
                    bp::make_setter(&OCPWalkParams::coneAxisWeight), ".")
      .add_property("refForceWeight", bp::make_getter(&OCPWalkParams::refForceWeight),
                    bp::make_setter(&OCPWalkParams::refForceWeight), ".")
      .add_property("impactAltitudeWeight", bp::make_getter(&OCPWalkParams::impactAltitudeWeight),
                    bp::make_setter(&OCPWalkParams::impactAltitudeWeight), ".")
      .add_property("impactVelocityWeight", bp::make_getter(&OCPWalkParams::impactVelocityWeight),
                    bp::make_setter(&OCPWalkParams::impactVelocityWeight), ".")
      .add_property("impactRotationWeight", bp::make_getter(&OCPWalkParams::impactRotationWeight),
                    bp::make_setter(&OCPWalkParams::impactRotationWeight), ".")
      .add_property("refMainJointsAtImpactWeight", bp::make_getter(&OCPWalkParams::refMainJointsAtImpactWeight),
                    bp::make_setter(&OCPWalkParams::refMainJointsAtImpactWeight), ".")
      .add_property("verticalFootVelWeight", bp::make_getter(&OCPWalkParams::verticalFootVelWeight),
                    bp::make_setter(&OCPWalkParams::verticalFootVelWeight), ".")
      .add_property("flyHighSlope", bp::make_getter(&OCPWalkParams::flyHighSlope),
                    bp::make_setter(&OCPWalkParams::flyHighSlope), ".")
      .add_property("flyHighWeight", bp::make_getter(&OCPWalkParams::flyHighWeight),
                    bp::make_setter(&OCPWalkParams::flyHighWeight), ".")
      .add_property("groundColWeight", bp::make_getter(&OCPWalkParams::groundColWeight),
                    bp::make_setter(&OCPWalkParams::groundColWeight), ".")
      .add_property("footMinimalDistance", bp::make_getter(&OCPWalkParams::footMinimalDistance),
                    bp::make_setter(&OCPWalkParams::footMinimalDistance), ".")
      .add_property("feetCollisionWeight", bp::make_getter(&OCPWalkParams::feetCollisionWeight),
                    bp::make_setter(&OCPWalkParams::feetCollisionWeight), ".")
      .add_property("kktDamping", bp::make_getter(&OCPWalkParams::kktDamping),
                    bp::make_setter(&OCPWalkParams::kktDamping), ".")
      .add_property("stateTerminalWeight", bp::make_getter(&OCPWalkParams::stateTerminalWeight),
                    bp::make_setter(&OCPWalkParams::stateTerminalWeight), ".")
      .add_property("solver_th_stop", bp::make_getter(&OCPWalkParams::solver_th_stop),
                    bp::make_setter(&OCPWalkParams::solver_th_stop), ".")
      .add_property("transitionDuration", bp::make_getter(&OCPWalkParams::transitionDuration),
                    bp::make_setter(&OCPWalkParams::transitionDuration), ".")
      .add_property("withNormalForceBoundOnly", bp::make_getter(&OCPWalkParams::withNormalForceBoundOnly),
                    bp::make_setter(&OCPWalkParams::withNormalForceBoundOnly), ".")
      .add_property("minimalNormalForce", bp::make_getter(&OCPWalkParams::minimalNormalForce),
                    bp::make_setter(&OCPWalkParams::minimalNormalForce), ".");
}

void exposeRefForce() {
  bp::def("computeWeightShareSmoothProfile", &computeWeightShareSmoothProfile,
          bp::args("contact_pattern", "duration", "saturation"),
          "Compute the smooth weight transfer between contact points");
}

void exposeOCPWalkclass() {
  bp::register_ptr_to_python<boost::shared_ptr<OCPWalk> >();

  bp::class_<OCPWalk>("OCPWalk",
                      bp::init<boost::shared_ptr<OCPRobotWrapper>, boost::shared_ptr<OCPWalkParams>, Eigen::MatrixXd>(
                          bp::args("self", "robot", "params", "contact_patter"),
                          "Initialize the OCP from robot, params and contact pattern"))
      // .add_property(
      //     "vcomRef",
      //     bp::make_getter(&OCPWalk::vcomRef,
      //     bp::return_internal_reference<>()),
      //     bp::make_setter(&OCPWalk::vcomRef),
      //     "Reference of the com velocity, to tune the OCP at runtime.")
      // .add_property("solver_th_stop",
      // bp::make_getter(&OCPWalk::solver_th_stop),
      //               bp::make_setter(&OCPWalk::solver_th_stop),
      //               "Stop threshold to configure the solver.")

      .def("buildRunningModels", &OCPWalk::buildRunningModels)

      .def("buildSolver", &OCPWalk::buildSolver)
      .def("buildInitialGuess", &OCPWalk::buildInitialGuess)
      .add_property("problem", bp::make_getter(&OCPWalk::problem, bp::return_value_policy<bp::return_by_value>()),
                    "Shooting problem used for OCP solver")
      .add_property("solver", bp::make_getter(&OCPWalk::solver, bp::return_value_policy<bp::return_by_value>()),
                    "Shooting solver")
      .add_property("state", bp::make_getter(&OCPWalk::state, bp::return_value_policy<bp::return_by_value>()),
                    "State model of the terminal node")
      .add_property("referenceForces", bp::make_getter(&OCPWalk::referenceForces),
                    bp::make_setter(&OCPWalk::referenceForces), ".");
}

void exposeOCPWalk() {
  exposeOCPRobotWrapper();
  exposeOCPParams();
  exposeOCPWalkclass();
  exposeRefForce();
}

}  // namespace python
}  // namespace sobec
